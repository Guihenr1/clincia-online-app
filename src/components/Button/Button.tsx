import { FC, MouseEvent, ReactNode } from "react";
import ButtonMaterialUI from "@mui/material/Button";
import useStyles from "./styles";
import { IconButton } from "@mui/material";

interface ButtonProps {
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  disabled?: boolean;
  children: ReactNode;
  variant?: "contained" | "outlined" | "text";
  isIconButton?: boolean;
  title?: string;
  color?:
    | "inherit"
    | "warning"
    | "primary"
    | "secondary"
    | "success"
    | "error"
    | "info";
  className?: string;
}

const Button: FC<ButtonProps> = ({
  onClick,
  disabled,
  children,
  variant,
  isIconButton,
  title,
  color,
  className,
}) => {
  const s = useStyles();

  const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick && !disabled) {
      onClick(event);
    }
  };

  return (
    <>
      {isIconButton ? (
        <IconButton
          className={s.classes.button}
          disabled={disabled}
          title={title}
          onClick={handleClick}
        >
          {children}
        </IconButton>
      ) : (
        <ButtonMaterialUI
          onClick={handleClick}
          disabled={disabled}
          variant={variant}
          className={`${s.classes.button} ${className}`}
          title={title}
          color={color}
        >
          {children}
        </ButtonMaterialUI>
      )}
    </>
  );
};

Button.defaultProps = {
  disabled: false,
  variant: "contained",
  isIconButton: false,
  title: "",
  color: "primary",
  className: "",
};

export default Button;
