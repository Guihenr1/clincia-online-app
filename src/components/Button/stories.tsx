import { MouseEvent } from "react";
import { Meta } from "@storybook/react";
import Button from "./Button";
import BugReportIcon from "@mui/icons-material/BugReport";

function onClick(e: MouseEvent<HTMLButtonElement>): void {
  console.log("CLICKED", e);
}

export default {
  component: Button,
  args: {
    onClick: onClick,
  },
} as Meta;

export const Default = (args: any) => <Button {...args} />;
export const Icon = (args: any) => <Button {...args} />;

Default.args = {
  children: "Button",
};

Icon.args = {
  children: <BugReportIcon />,
  title: "Bug Report",
  isIconButton: true,
};
