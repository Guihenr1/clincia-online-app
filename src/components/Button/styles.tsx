import { makeStyles } from "tss-react/mui";

const useStyles = makeStyles()(() => ({
  button: {},
}));

export default useStyles;
