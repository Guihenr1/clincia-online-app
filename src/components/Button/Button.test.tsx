import { fireEvent, render, screen } from "@testing-library/react";
import Button from "./Button";
import "@testing-library/jest-dom/extend-expect";

describe("Button", () => {
  test("renders the button", () => {
    const buttonText = "Click me";
    const { getByText } = render(<Button>{buttonText}</Button>);

    expect(screen.getByRole("button")).toHaveTextContent(/Click me/);
    const buttonElement = getByText(buttonText);
    expect(buttonElement).toBeInTheDocument();
  });

  test("calls onClick function when clicked", () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <Button onClick={onClickMock}>Click me</Button>
    );
    const buttonElement = getByText("Click me");
    fireEvent.click(buttonElement);
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test("does not call onClick function when clicked", () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <Button disabled onClick={onClickMock}>
        Click me
      </Button>
    );
    const buttonElement = getByText("Click me");
    fireEvent.click(buttonElement);
    expect(onClickMock).not.toHaveBeenCalled();
  });

  test("must apply the default variant", () => {
    const buttonText = "Click me";
    render(<Button>{buttonText}</Button>);

    expect(screen.getByRole("button")).toHaveClass("MuiButton-contained");
  });

  test("must apply the correct variant", () => {
    const variants = ["contained", "outlined", "text"];

    variants.forEach((variant) => {
      render(
        <Button variant={variant as "contained" | "outlined" | "text"}>
          {variant}
        </Button>
      );

      expect(screen.getByText(`${variant}`)).toHaveClass(
        `MuiButton-${variant}`
      );
    });
  });
});
