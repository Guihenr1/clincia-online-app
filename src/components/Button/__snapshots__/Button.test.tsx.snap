// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders the button 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1wwchiu-MuiButtonBase-root-MuiButton-root-button"
    tabindex="0"
    type="button"
  >
    Click me
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
